numba/core/image_gen.py

```diff
@@ -0,0 +1,480 @@
+
+        cheap_name = "Module passes (cheap optimization for refprune)"
+        with self._recorded_timings.record(cheap_name):
+            # A cheaper optimisation pass is run first to try and get as many
+            # refops into the same function as possible via inlining
+            self._codegen._mpm_cheap.run(mod)
+        # Refop pruning is then run on the heavily inlined function
+        if not config.LLVM_REFPRUNE_PASS:
+            mod = remove_redundant_nrt_refct(mod)
+        full_name = "Module passes (full optimization)"
+        with self._recorded_timings.record(full_name):
+            # The full optimisation suite is then run on the refop pruned IR
+            self._codegen._mpm_full.run(mod)
+
+        mod.verify()
+
+        # Scan for dynamic globals
+        for gv in mod.global_variables:
+            if gv.name.startswith('numba.dynamic.globals'):
+                self._dynamic_globals.append(gv.name)
+
+        self._native_asm = self._codegen._target_machine.emit_assembly(mod)
+        object_code = self._codegen._target_machine.emit_object(mod)
+
+        if config.DUMP_OPTIMIZED:
+            dump("OPTIMIZED DUMP %s" % self.name,
+                 '\n'.join(self.get_llvm_str()),
+                 'llvm')
+
+        if config.DUMP_ASSEMBLY:
+            dump("ASSEMBLY %s" % self.name, self.get_asm_str(), 'asm')
+
+        self._materialization_result = self._codegen._handle_image(
+            self.name,
+            self.unique_name,
+            object_code,
+            self._native_asm,
+            self._linking_libraries,
+            self._globals)
+
+    def get_defined_functions(self):
+        """
+        Get all functions defined in the library.  The library must have
+        been finalized.
+        """
+        self._ensure_finalized()
+        for fn, fn_ty in self._globals:
+            yield llvmlite.ir.Function(fn, None, fn_ty)
+
+    def get_function(self, name):
+        ty = self._globals[name]
+        if isinstance(ty, llvmir.FunctionType):
+            return llvmlite.ir.Function(name, None, ty)
+
+    def get_llvm_str(self):
+        return [*self._llvm_ir]
+
+    def get_asm_str(self):
+        return self._native_asm
+
+
+ImageConfiguration = namedtuple('ImageConfiguration',
+                                ('triple', 'cpu_name', 'features'))
+
+
+class BaseImageCodegen(Codegen):
+
+    def __init__(self, image_configuration):
+        super().__init__()
+        initialize_llvm()
+        target = ll.Target.from_triple(image_configuration.triple)
+        tm_options = dict(opt=config.OPT)
+        if image_configuration.cpu_name == 'host':
+            cpu_name = (ll.get_host_cpu_name()
+                        if config.CPU_NAME is None
+                        else config.CPU_NAME)
+        else:
+            cpu_name = image_configuration.cpu_name
+        tm_options['cpu'] = cpu_name
+        tm_options['reloc'] = 'pic'
+        tm_options['codemodel'] = 'default'
```

Comment:
```suggestion
        tm_options['codemodel'] = 'jitdefault'
```

On the M1 Mac, the default code model ended up as "small" and causes failure in OrcJIT to find in-process symbols (e.g. `puts`). I made a reproducer at https://gist.github.com/sklam/108e89c745de9d9cbc4fa5154721d760, which tries to use `puts` (optimized from `printf`) and `lldb` shows generated machine code binds to the wrong address. The program will segfault.

Searching for the problem led me to https://github.com/llvm/llvm-project/issues/84#issuecomment-576900826 which suggested using the "jitdefault" codemodel. I verified this change worked on my M1 Mac. But we need to verify how portable this really is on different OS+arch given past experience with MCJIT.